examples/bsoncxx/builder_core.cpp

```diff
@@ -0,0 +1,55 @@
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/types.hpp>
+
+using namespace bsoncxx;
+
+int main(int, char**) {
+
+    // bsoncxx::builder::core is a low-level primitive that can be useful for building other
+    // BSON abstractions. Most users should just use builder::stream or builder::basic.
+
+    // the boolean argument to core should be true if the top-level BSON datum is an array.
+    // if false, it will be created as a document.
+    auto builder = builder::core{false};
+
+    // When building a document, we have two ways of passing keys.
+
+    // 1. We can pass a key using key_owned(). This is less efficient but
+    // frees us from worrying about the lifetime of the key.
+
+    // Example: if we have a stack allocated std::string, key_owned will take ownership
+    // of it, so the following code is legal.
+    {
+        std::string temp{"foo"};
+        builder.key_owned(temp);
+        builder.append(types::b_bool{false});
+    }
+
+    // 2. We can pass a key using key_view(). This is more efficient (less copying)
+    // but we need to manage the lifetime of the key until a matching value is appended, at which
+    // point it will be copied into an internally managed buffer.
+
+    {
+        // THIS IS ILLEGAL: DO NOT DO THIS
```

Comment:
the builder itself goes out of scope here at the same time it becomes illegal to use, which should hopefully prevent ASAN from complaining.